### PR TITLE
Optimize regexp matcher with list of alternations ending with .*

### PR DESCRIPTION
**Pull request by pracucci**
While investigating some query performance issues in production, I noticed that the set of very slow queries all had non-optimized regexp matchers in the format `(?i)(a.*|b.*|c.*)` (with many more alternations following the same pattern). In this PR I propose to optimize for such case too.

The proposed change is intentionally pretty small, in order to make it easier to review and reduce risks.

Analyzing the customer queries, here the difference in the number of optimized customer queries (out of 72018 queries analyzed):
- Before: 71.21%
- After: 99.39%

What the benchmark says (the 2nd last entry is the customer regexp). I've omitted memory comparison because there's no difference:

```
goos: darwin
goarch: amd64
pkg: github.com/prometheus/prometheus/model/labels
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
                                                     │  before.txt   │              after.txt              │
                                                     │    sec/op     │   sec/op     vs base                │
FastRegexMatcher/#00-12                                  61.81n ± 4%   62.21n ± 2%        ~ (p=0.896 n=10)
FastRegexMatcher/foo-12                                  76.14n ± 4%   74.79n ± 7%        ~ (p=0.143 n=10)
FastRegexMatcher/^foo-12                                 61.77n ± 3%   61.07n ± 2%        ~ (p=0.436 n=10)
FastRegexMatcher/(foo|bar)-12                            66.71n ± 3%   64.46n ± 3%   -3.38% (p=0.011 n=10)
FastRegexMatcher/foo.*-12                                147.6n ± 2%   139.2n ± 3%   -5.63% (p=0.000 n=10)
FastRegexMatcher/.*foo-12                                169.2n ± 3%   173.7n ± 3%        ~ (p=0.085 n=10)
FastRegexMatcher/^.*foo$-12                              170.4n ± 3%   173.6n ± 3%        ~ (p=0.138 n=10)
FastRegexMatcher/^.+foo$-12                              168.0n ± 5%   171.2n ± 3%        ~ (p=0.184 n=10)
FastRegexMatcher/.*-12                                   131.3n ± 5%   127.0n ± 3%        ~ (p=0.089 n=10)
FastRegexMatcher/.+-12                                   127.3n ± 5%   127.0n ± 5%        ~ (p=0.224 n=10)
FastRegexMatcher/foo.+-12                                148.4n ± 3%   139.2n ± 5%   -6.23% (p=0.000 n=10)
FastRegexMatcher/.+foo-12                                166.6n ± 3%   168.4n ± 2%        ~ (p=0.493 n=10)
FastRegexMatcher/foo_.+-12                               134.9n ± 5%   133.1n ± 4%        ~ (p=0.383 n=10)
FastRegexMatcher/foo_.*-12                               134.8n ± 4%   137.8n ± 4%        ~ (p=0.393 n=10)
FastRegexMatcher/.*foo.*-12                              280.9n ± 4%   284.3n ± 4%        ~ (p=0.325 n=10)
FastRegexMatcher/.+foo.+-12                              293.9n ± 4%   295.2n ± 3%        ~ (p=0.871 n=10)
FastRegexMatcher/(?s:.*)-12                              58.61n ± 5%   58.98n ± 3%        ~ (p=0.565 n=10)
FastRegexMatcher/(?s:.+)-12                              73.86n ± 6%   71.20n ± 3%   -3.59% (p=0.000 n=10)
FastRegexMatcher/(?s:^.*foo$)-12                         168.6n ± 3%   162.2n ± 5%        ~ (p=0.101 n=10)
FastRegexMatcher/(?i:foo)-12                             112.7n ± 3%   113.4n ± 5%        ~ (p=0.725 n=10)
FastRegexMatcher/(?i:(foo|bar))-12                       240.4n ± 8%   243.8n ± 4%        ~ (p=0.739 n=10)
FastRegexMatcher/(?i:(foo1|foo2|bar))-12                 426.8n ± 6%   418.3n ± 6%   -1.99% (p=0.043 n=10)
FastRegexMatcher/^(?i:foo|oo)|(bar)$-12                  1.062µ ± 4%   1.054µ ± 5%        ~ (p=0.684 n=10)
FastRegexMatcher/(?i:(foo1|foo2|aaa|bbb|ccc|ddd|e-12     1.889µ ± 1%   1.910µ ± 1%   +1.11% (p=0.000 n=10)
FastRegexMatcher/((.*)(bar|b|buzz)(.+)|foo)$-12          688.4n ± 4%   667.9n ± 3%        ~ (p=0.075 n=10)
FastRegexMatcher/^$-12                                   63.50n ± 4%   64.28n ± 3%        ~ (p=0.565 n=10)
FastRegexMatcher/(prometheus|api_prom)_api_v1_.+-12      226.9n ± 3%   230.1n ± 6%        ~ (p=0.247 n=10)
FastRegexMatcher/10\.0\.(1|2)\.+-12                      134.8n ± 4%   136.1n ± 3%        ~ (p=0.698 n=10)
FastRegexMatcher/10\.0\.(1|2).+-12                       136.7n ± 3%   134.2n ± 2%        ~ (p=0.148 n=10)
FastRegexMatcher/((fo(bar))|.+foo)-12                    311.3n ± 7%   284.2n ± 4%   -8.69% (p=0.000 n=10)
FastRegexMatcher/zQPbMkNO|NNSPdvMi|iWuuSoAl|qbvKM-12     289.5n ± 5%   296.2n ± 4%        ~ (p=0.515 n=10)
FastRegexMatcher/jyyfj00j0061|jyyfj00j0062|jyyfj9-12     314.2n ± 6%   315.5n ± 3%        ~ (p=0.796 n=10)
FastRegexMatcher/(?i:(zQPbMkNO|NNSPdvMi|iWuuSoAl|-12     1.898µ ± 1%   1.909µ ± 1%   +0.55% (p=0.006 n=10)
FastRegexMatcher/(?i:(zQPbMkNO.*|NNSPdvMi.*|iWuuS-12    46.267µ ± 4%   9.513µ ± 4%  -79.44% (p=0.000 n=10)
FastRegexMatcher/(?i:(.*zQPbMkNO|.*NNSPdvMi|.*iWu-12   1549.88µ ± 5%   11.38µ ± 5%  -99.27% (p=0.000 n=10)
geomean                                                  286.7n        236.4n       -17.53%
```

I've also run the fuzzy tests for a while:

```
$ go test -run '^$' -fuzz 'FuzzFastRegexMatcher_WithStaticallyDefinedRegularExpressions' ./model/labels

fuzz: elapsed: 0s, gathering baseline coverage: 0/312 completedfuzz: elapsed: 1s, gathering baseline coverage: 312/312 completed, now fuzzing with 12 workers
fuzz: elapsed: 3s, execs: 23161 (7719/sec), new interesting: 9 (total: 321)
fuzz: elapsed: 6s, execs: 48026 (8288/sec), new interesting: 11 (total: 323)
fuzz: elapsed: 9s, execs: 71033 (7668/sec), new interesting: 15 (total: 327)
...
fuzz: elapsed: 5m39s, execs: 2648543 (13497/sec), new interesting: 85 (total: 397)
```

```
$ go test -run '^$' -fuzz 'FuzzFastRegexMatcher_WithFuzzyRegularExpressions' ./model/labels

fuzz: elapsed: 1s, gathering baseline coverage: 0/3156 completed
fuzz: elapsed: 4s, gathering baseline coverage: 1269/3156 completed
fuzz: elapsed: 7s, gathering baseline coverage: 2580/3156 completed
fuzz: elapsed: 8s, gathering baseline coverage: 3156/3156 completed, now fuzzing with 12 workers
fuzz: elapsed: 10s, execs: 20376 (5940/sec), new interesting: 3 (total: 3159)
fuzz: elapsed: 13s, execs: 49398 (9675/sec), new interesting: 10 (total: 3166)
fuzz: elapsed: 16s, execs: 50765 (456/sec), new interesting: 11 (total: 3167)
...
fuzz: elapsed: 2m52s, execs: 367428 (5230/sec), new interesting: 54 (total: 3210)
```

### CODE REVIEW

File: model/labels/regexp_test.go [hunk 1]
```diff
@@ -400,10 +404,15 @@ func Test_OptimizeRegex(t *testing.T) {
 		{".*foo.*bar.*", nil},
 		{`\d*`, nil},
 		{".", nil},
+		{"/|/bar.*", &literalPrefixStringMatcher{prefix: "/", prefixCaseSensitive: true, right: orStringMatcher{emptyStringMatcher{}, &literalPrefixStringMatcher{prefix: "bar", prefixCaseSensitive: true, right: anyStringWithoutNewlineMatcher{}}}}},
 		// This one is not supported because  `stringMatcherFromRegexp` is not reentrant for syntax.OpConcat.
 		// It would make the code too complex to handle it.
-		{"/|/bar.*", nil},
```
Comment (pracucci) [inline model/labels/regexp_test.go:405]: Accidentally removed. Going to re-introduce it.

Comment (pracucci) [inline model/labels/regexp_test.go:405]: Not removed, but changed above (because now it's optimised).

File: model/labels/regexp_test.go [hunk 2]
```diff
@@ -498,7 +665,7 @@ func TestAnalyzeRealQueries(t *testing.T) {
 		queryStartTimes []time.Time
 	}
 
-	labelValueRE := regexp.MustCompile(`=~\\"([^"]+)\\"`)
+	labelValueRE := regexp.MustCompile(`=~(?:\\"|')([^"']*)(?:\\"|')`)
```
Comment (pracucci) [inline model/labels/regexp_test.go:668]: Changed to also match single quotes (which I've just learned they're supported and customer is using in few queries).

**Review by colega — APPROVED** [review]
LGTM. This took me a while to get my head around this.

**Comment by pracucci** [comment]
> LGTM. This took me a while to get my head around this.

This stuff is very tricky 🤞

File: model/labels/regexp.go [hunk 3]
```diff
@@ -524,11 +527,39 @@ func stringMatcherFromRegexpInternal(re *syntax.Regexp) StringMatcher {
...
+		}
+
+		// Ensure we've found some literals to match (optionally with a left and/or right matcher).
+		// If not, then this optimization doesn't trigger.
 		if len(matches) == 0 {
 			return nil
 		}
 
-		if left == nil && right == nil {
+		// Use the right (and best) matcher based on what we've found.
```
Comment (bboreham) [inline model/labels/regexp.go:559]: Does "right" mean "opposite of left" or "correct" here?
Might be worth rewording.

File: model/labels/regexp.go [hunk 4]
```diff
@@ -524,11 +527,39 @@ func stringMatcherFromRegexpInternal(re *syntax.Regexp) StringMatcher {
 		}
 
 		matches, matchesCaseSensitive := findSetMatchesInternal(re, "")
+
+		if len(matches) == 0 && len(re.Sub) == 2 {
+			// We have not find fixed set matches. We look for other known cases that
```
Comment (bboreham) [inline model/labels/regexp.go:532]:
```suggestion
			// We have not found a fixed set of matches. We look for other known cases that
```

**Review by bboreham — COMMENTED** [review]
I don't see anything wrong with the code, but I do find it very difficult to follow.

**Pull request merged**